batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+	}
+	log.Println("Read done")
+}
+
+func (p *Pipeline) Download(wg *sync.WaitGroup) {
+	// Download the image from the URL
+	// Send the image to the convert channel
+	defer wg.Done()
+	var (
+		row  *Out
+		ok   bool
+		body io.Reader
+		hash string
+		ext  string
+		err  error
+	)
+	for {
+
+		row, ok = <-p.channels[READ]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		// Check if the URL has already been processed
+		// If it has, skip it
+		_, ok := p.urlMap.Load(row.Url)
+		if ok {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: ErrDuplicateURL,
+			}
+			continue
+		} else {
+			p.urlMap.Store(row.Url, true)
+		}
+
+		body, ext, hash, err = DownloadWithBackoff(row.Url, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		key := fmt.Sprintf("%s-converted.%s", hash, ext)
+
+		// Check if the image hash is already in the S3 bucket
+		_, err = p.config.Aws.GetObject(&s3.GetObjectInput{
+			Bucket: &p.config.Aws.s3bucket,
+			Key:    &key,
+		})
+
+		if err == nil {
+			row.Err = ErrImageExistsInS3
+			row.S3url = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", p.config.Aws.s3bucket, hash)
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		//Create an empty file
+		inputPath := fmt.Sprintf("/outputs/%s.%s", hash, ext)
+		outputPath := fmt.Sprintf("/outputs/%s", key)
+
+		file, err := os.Create(inputPath)
+		if err != nil {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: err,
+			}
+			continue
+		}
+		defer file.Close()
+
+		//Write bytes to the file
+		_, err = io.Copy(file, body)
+		if err != nil {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: err,
+			}
+			continue
+		}
+
+		p.channels[DOWNLOAD] <- &Out{
+			Url:    row.Url,
+			Input:  inputPath,
+			Output: outputPath,
+		}
+	}
+}
+
+func (p *Pipeline) Convert(wg *sync.WaitGroup) {
+	// Convert the image
+	// Send the image to the upload channel
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[DOWNLOAD]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[CONVERT] <- &Out{
+				Url:    row.Url,
+				Err:    row.Err,
+				Input:  row.Input,
+				Output: "",
+			}
+			continue
+		}
+		err := p.config.Converter.Grayscale(row.Input, row.Output)
+		if err != nil {
+			p.channels[CONVERT] <- &Out{
+				Url:    row.Url,
+				Err:    err,
+				Input:  row.Input,
+				Output: "",
+			}
+			continue
+		}
+		p.channels[CONVERT] <- row
+	}
+}
+
+func (p *Pipeline) Upload(wg *sync.WaitGroup) {
+	// Upload the image to S3
+	// Send the image to the result channel
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[CONVERT]
+		if !ok {
+			break
+		}
+
+		if row.Err != nil {
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		file, err := os.Open(row.Output)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		// Upload to S3
+		path := strings.Split(row.Output, "/")
```

Comment:
It feels like `row.Output` should probably have some documentation on it as to what format it's in (so that the caller here knows it's safe to split on `/` and take the last component), and possibly a method on `Out` to get the `S3BaseName` rather than have the caller construct it.

I'd even go a step further and say: `Output` is maybe a strange thing to save in the struct at all - I'd maybe save the `hash` and `fileType`, and construct the S3 path from those in a method, rather than smooshing them all together in a field and then separating them out again here...

Reply:
Makes sense, has been applied now. Much cleaner, thanks!